Lib/pdb.py

```diff
@@ -590,6 +591,36 @@ def default(self, line):
         except:
             self._error_exc()
 
+    def _replace_convenience_variable(self, line):
+        """Replace the convenience variable in line"""
+        last_token_is_dollar = False
+        dollar_start = None
+        dollar_end = None
+        replace_variables = []
+        try:
+            for t in tokenize.generate_tokens(io.StringIO(line).readline):
+                token_type, token_string, start, end, _ = t
+                if token_type == token.OP and token_string == '$':
+                    last_token_is_dollar = True
+                    dollar_end = end
+                    dollar_start = start
+                else:
+                    if (last_token_is_dollar and
+                        token_type == token.NAME and
+                        start == dollar_end):
+                        # line is a one line command so we only care about column
+                        replace_variables.append((dollar_start[1], end[1], token_string))
+                    last_token_is_dollar = False
+        except tokenize.TokenError:
+            return line
+        last_end = 0
+        new_line = ''
+        for start, end, name in replace_variables:
+            new_line += line[last_end:start] + f'__pdb_convenience_variables["{name}"]'
+            last_end = end
+        new_line += line[last_end:]
+        return new_line
```

Comment:
Instead of repeatedly creating strings with `+=`, it can be cleaner to use an array for the `new_line` parts, and join them when returning.

Reply:
We should add extra breaks in more common cases to avoid unnecessary computation like you mentioned in the other comments. I'll fix those soon. However, I don't think concatenating strings is less clean than joining an array - the code is almost exactly the same, except for an extra line of `join` at the end.

Joining a list is often preferred because it's faster and more memory friendly than repeatedly concatenating strings, but in the actual use case, it would be very rare to have multiple convenience variables to replace. So the gain for using lists instead of string concatenation is not significant. String concatenation is clearer semantically in my opinion.

Reply:
> String concatenation is clearer semantically in my opinion.

Im not sure, string concatenation in a loop jumps out as a code smell.